### PR TITLE
8263185: Mallinfo deprecated in glibc 2.33

### PR DESCRIPTION
**Pull request by mychris**
Starting with glibc 2.33, mallinfo is deprecated in favor of the new mallinfo2 API. Both APIs work the same way, the only difference is, that mallinfo2 uses size_t instead of int for the fields of the struct, containing the information about malloc. Please see the [glibc release notes](https://sourceware.org/pipermail/libc-alpha/2021-February/122207.html).

Testing with tier1 on a system with glibc 2.33 and on a system with glibc 2.31.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263185](https://bugs.openjdk.java.net/browse/JDK-8263185): Mallinfo deprecated in glibc 2.33

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2964/head:pull/2964`
`$ git checkout pull/2964`

To update a local copy of the PR:
`$ git checkout pull/2964`
`$ git pull https://git.openjdk.java.net/jdk pull/2964/head`

### CODE REVIEW

**Comment by tstuefe** [comment]
Would this not mean we fail to run a glibc 2.33+ built VM on a machine with an older glibc?

**Comment by mychris** [comment]
Yes. I was thinking about using dlsym to detect at runtime which functions are available.
But because of symbol versioning, I expect the runtime linker to fail linking the executable, if it requires a glibc 2.33 and the system only has a glibc with a lower version. This is already the case.
Or do you expect the linker to ignore symbol versioning? Are we supporting linux platforms which are not using it?

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -2229,19 +2229,24 @@ void os::Linux::print_process_memory_info(outputStream* st) {
   // Print glibc outstanding allocations.
   // (note: there is no implementation of mallinfo for muslc)
 #ifdef __GLIBC__
+#if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
+  struct mallinfo2 mi = ::mallinfo2();
+  const size_t total_allocated = mi.uordblks;
+  const bool might_have_wrapped = false;
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:2235]: Can you move declaration of might_have_wrapped - defaulted to false - outside the define? Same for total_allocated.

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -2229,19 +2229,24 @@ void os::Linux::print_process_memory_info(outputStream* st) {
...
+#else // __GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ < 33)
   struct mallinfo mi = ::mallinfo();
-
   // mallinfo is an old API. Member names mean next to nothing and, beyond that, are int.
   // So values may have wrapped around. Still useful enough to see how much glibc thinks
   // we allocated.
-  const size_t total_allocated = (size_t)(unsigned)mi.uordblks;
-  st->print("C-Heap outstanding allocations: " SIZE_FORMAT "K", total_allocated / K);
+  const size_t total_allocated = static_cast<size_t>(mi.uordblks);
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:2241]: The original version was correct here, please restore it. You need to make the int value unsigned before extending it to 64bit size_t, otherwise you get sign extension for values >= 2g. 

One could argue that the second cast to size_t in the original coding was superfluous, but the cast to unsigned was needed.

File: src/hotspot/os/linux/os_linux.cpp [hunk 3]
```diff
@@ -2229,19 +2229,24 @@ void os::Linux::print_process_memory_info(outputStream* st) {
...
   // Since mallinfo members are int, glibc values may have wrapped. Warn about this.
-  if ((vmrss * K) > UINT_MAX && (vmrss * K) > (total_allocated + UINT_MAX)) {
+  const bool might_have_wrapped = (vmrss * K) > UINT_MAX && (vmrss * K) > (total_allocated + UINT_MAX);
+#endif // __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
+  st->print("C-Heap outstanding allocations: " SIZE_FORMAT "K", total_allocated / K);
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:2245]: Shorten this to:
```
st->print("C-Heap outstanding allocations: " SIZE_FORMAT "K%s",
           total_allocated / K, wrapped ? " (possibly wrapped)" : "");
```
?

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi Chris,

are you sure this is true? That we are unable to run on machines with older glibcs than the build machine uses? This is new to me. 

If this is not true, I would dynamically resolve mallinfo2. This is too unimportant a functionality to introduce a hard glibc dependency.

Minor remarks inline.

..Thomas

**Comment by tstuefe** [comment]
Also please make sure you test on both new and old glibc and run some manual sanity tests. E.g. by connecting to a running VM via jcmd and executing VM.info. The output should appear somewhere in there, and match your expectations. The sign extension issue should have appeared when running on an older glibc with 50% probability as a unseemingly high value.

**Comment by mychris** [comment]
Thanks for your review, Thomas.

> are you sure this is true? That we are unable to run on machines with older glibcs than the build machine uses? This is new to me.

That's how I understand and experience symbol versioning. But I don't know if there should be support for Linux using glibc without symbol versioning. Also, I don't know in detail when the glibc bumps a symbol version, so it might be better to not rely on it here. I changed the implementation to use `dlsym` to resolve the symbols without any specific version.

I am resolving both functions, `mallinfo2` and `mallinfo`, because I am using `mallinfo` as a fallback function if we compiled against glibc >= 2.33. If `mallinfo` is not resolved using `dlsym`, there will be a compile time warning because of its deprecation. This compile time warning was the reason I opened up the bug report.

I fixed the regression I introduced and double checked that the overflow is handled as before. As you mentioned, the VM doesn't allocate enough under normal circumstances, so I didn't notice this during testing. I added additional allocations as suggested and tested again. Works as before now. I tested:

* build glibc 2.31 - runtime glibc 2.31
* build glibc 2.33 - runtime glibc 2.33
* build glibc 2.31 - runtime glibc 2.33

Output is as before, only that the VM built with 2.33 doesn't have the overflow problem anymore.

I also executed tier1 on both systems.

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi Christoph,

this looks fine in principle, but could you please move the symbol resolution to the init phase? There are plenty of examples, e.g. `os::Linux::sched_getcpu_init`, `os::Linux::libnuma_init`. (They are all exposed via the os::Linux namespace which I personally don't like but when in Rome...)

The main reason for this is that `os::Linux::print_process_memory_information` may be called during error reporting, in the context of a crash, and I don't want to call as little system APIs as possible there.

BTW, you can also test this with -XX:ErrorHandlerTest=*number*, with e.g. 14 I believe giving you a real SEGV. The information should be printed correctly as part of the hs-err file.

Also please use CAST_TO_FN_PTR instead of manually casting. In general, its good to keep to the existing style wrt to C++ features. For one, we have a list of allowed and disallowed list of C++ features, which is not immediately clear (see hotspot style guide in top level doc folder). And I personally don't like mixing styles, better to change code to more modern C++ syntax in specific RFEs if that is wanted. Also makes backporting functional changes easier.

Thanks, Thomas

**Comment by mychris** [comment]
Hi Thomas,

> Also please use CAST_TO_FN_PTR instead of manually casting.

Sure thing, changed in the last commit.

About moving the symbol resolution. Before changing this, I would like to get your feedback, because the reason I didn't do it where you are suggesting was the amount of GLIBC version dependent code I would introduce. If I move the the resolution (and keep the current logic), I would add 3 more `ifdefs` which check the glibc version:

1. Delcaration of the two function pointers
2. Initialization of the two function pointer
3. Symbol resolution

This seemed a lot of dependent code, all of which need two `#if`:

```
#ifdef GLIBC
// declare (os_linux.hpp)  / intialize (os_linux.cpp) / resolve (os_linux.cpp os::init) mallinfo function pointer
#if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
// declare / intialize / resolve mallinfo2 function pointer
#endif
#endif
```

We always need two pointers, since we need to know which one we are calling because of the return type. We could only have one and do something like the following, but then we loose the ability to call `mallinfo`, if the JVM has been linked against glibc >= 2.33:

```
#ifdef GLIBC
#if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
// declare / intialize / resolve mallinfo2 function pointer
#else
// declare / intialize / resolve mallinfo function pointer
#endif
#endif
```

I wanted to avoid that, that's why I did the resolution where it is right now.

We could also ignore `mallinfo2` for the moment, and always call `mallinfo`, regardless of the glibc version. I think the important part right now is, two get rid of the compile time warning (by resolving `mallinfo` at runtime using `dlsym`), not enhancing the error reporting.

**Comment by tstuefe** [comment]
> Hi Thomas,
> 
> > Also please use CAST_TO_FN_PTR instead of manually casting.
> 
> Sure thing, changed in the last commit.
> 
> About moving the symbol resolution. Before changing this, I would like to get your feedback, because the reason I didn't do it where you are suggesting was the amount of GLIBC version dependent code I would introduce. If I move the the resolution (and keep the current logic), I would add 3 more `ifdefs` which check the glibc version:
> 
> 1. Delcaration of the two function pointers
> 2. Initialization of the two function pointer
> 3. Symbol resolution
> 
> This seemed a lot of dependent code, all of which need two `#if`:
> 
> ```
> #ifdef GLIBC
> // declare (os_linux.hpp)  / intialize (os_linux.cpp) / resolve (os_linux.cpp os::init) mallinfo function pointer
> #if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
> // declare / intialize / resolve mallinfo2 function pointer
> #endif
> #endif
> ```
> 
> We always need two pointers, since we need to know which one we are calling because of the return type. We could only have one and do something like the following, but then we loose the ability to call `mallinfo`, if the JVM has been linked against glibc >= 2.33:
> 
> ```
> #ifdef GLIBC
> #if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
> // declare / intialize / resolve mallinfo2 function pointer
> #else
> // declare / intialize / resolve mallinfo function pointer
> #endif
> #endif
> ```
> 
> I wanted to avoid that, that's why I did the resolution where it is right now.
> 
> We could also ignore `mallinfo2` for the moment, and always call `mallinfo`, regardless of the glibc version. I think the important part right now is, two get rid of the compile time warning (by resolving `mallinfo` at runtime using `dlsym`), not enhancing the error reporting.

The code does not have to be complicated. You do not need any ifdefs. In fact, it would be better if you would avoid them.

1) locally define struct mallinfo2. Actually all we need is the correct offset for urordblks so in principle all other members could be called dummy1, dummy2...
2) define mallinfo2* (mallinfo2_fun)() = NULL;
3) in init, resolve mallinfo2
4) when reporting, if mallinfo2_fun=NULL, call it, otherwise call (statically linked) mallinfo. 

That way you don't have any #ifs; a VM compiled on an older machine will still show correct values when running on a machine with glibc2.33 and later. VM would be back-and-forward compatible. Patch should be quite small. 

You could even do completely without defining mallinfo2 structure:
1) define (void*)(mallinfo2_fun)() = NULL
2) resolve it
3) use it like this:

```
if (mallinfo2_fun != NULL) {
	void* p = mallinfo2_fun();
	size_t total_allocated = ((size_t*)p)[7];
}
```

which is perfectly fine since mallinfo2 is just a sequence of size_t values. Bit less code than the first alternative; the first alternative is a bit cleaner and more obvious.

All that said, I read David proposed to just mute the warning. I'm fine with that too though I would prefer to use mallinfo2. I leave it up to you.

**Comment by mychris** [comment]
Let's give it another shot or two, hopefully we find something good enough before simply disabling the warning. There is a gcc option to disable deprecation warnings, but this is only global AFAIK. There might also be a pragma which could be used for this function, but I have to see if it is supported in clang as well.

I don't like the idea of copying the struct definition from a standard header, or indexing it "blindly". This shouldn't be necessary for such a simple API. I came up with the following, which for me only has the downside, that we are always resolving `mallinfo` and `mallinfo2`, regardless of the glibc version, which is overhead during startup, but it shouldn't be too much of a deal. Also, the cast of the function pointer is done when it is used, which is not so nice, but we get around a couple of `#if` that way.

Tested with glibc 2.33 and 2.31.

File: src/hotspot/os/linux/os_linux.cpp [hunk 4]
```diff
@@ -2229,19 +2232,31 @@ void os::Linux::print_process_memory_info(outputStream* st) {
...
-  // we allocated.
-  const size_t total_allocated = (size_t)(unsigned)mi.uordblks;
-  st->print("C-Heap outstanding allocations: " SIZE_FORMAT "K", total_allocated / K);
-  // Since mallinfo members are int, glibc values may have wrapped. Warn about this.
-  if ((vmrss * K) > UINT_MAX && (vmrss * K) > (total_allocated + UINT_MAX)) {
-    st->print(" (may have wrapped)");
+  size_t total_allocated = 0;
+  bool called_mallinfo = false;
+  bool might_have_wrapped = false;
+#if __GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 33)
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:2238]: This is not needed - you are using a build-time value to determine whether to perform a runtime action. The mallinfo2 call should be attempted on all platforms regardless of the build platform characteristic.

File: src/hotspot/os/linux/os_linux.cpp [hunk 5]
```diff
@@ -170,6 +170,9 @@ const char * os::Linux::_libc_version = NULL;
 const char * os::Linux::_libpthread_version = NULL;
 size_t os::Linux::_default_large_page_size = 0;
 
+void* os::Linux::_mallinfo_fun_ptr = NULL;
+void* os::Linux::_mallinfo2_fun_ptr = NULL;
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:174]: Should this be ifdef __GLIBC__ too?

File: src/hotspot/os/linux/os_linux.cpp [hunk 6]
```diff
@@ -4406,6 +4421,9 @@ void os::init(void) {
 
   Linux::initialize_system_info();
 
+  Linux::_mallinfo_fun_ptr = dlsym(RTLD_DEFAULT, "mallinfo");
+  Linux::_mallinfo2_fun_ptr = dlsym(RTLD_DEFAULT, "mallinfo2");
+
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.cpp:4426]: Should this be ifdef __GLIBC__ too?

File: src/hotspot/os/linux/os_linux.hpp [hunk 7]
```diff
@@ -253,6 +253,9 @@ class Linux {
   };
   static NumaAllocationPolicy _current_numa_policy;
 
+  static void* _mallinfo_fun_ptr;
+  static void* _mallinfo2_fun_ptr;
```
Comment (dholmes-ora) [inline src/hotspot/os/linux/os_linux.hpp:257]: Should this be ifdef __GLIBC__ too?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Christoph,

There is no need for build-time glibc version checks - see below.

Thanks,
David

**Comment by tstuefe** [comment]
As I wrote before, I don't see a reason to resolve mallinfo dynamically.

File: src/hotspot/os/linux/os_linux.cpp (hunk 4 above)
Comment (mychris) [reply src/hotspot/os/linux/os_linux.cpp:2238]: But we need to know the structure of the return type, which needs to be known at compile time because it is allocated on the stack. The only way I see to make this not compile time dependent would be to copy the structure from the system header to the HotSpot sources and always assume that every platform, which provides a mallinfo2 function, has the same prototype. I would like to avoid copying the struct definition from the system header to the HotSpot sources. Or am I missing something?

**Comment by mychris** [comment]
>> As I wrote before, I don't see a reason to resolve mallinfo dynamically.
> 
> Doesn't statically using mallinfo trigger the warning that we are trying to avoid?

Yes, that's why I resolve it dynamically. Otherwise we would need to use a GCC pragma to disable deprecation warnings locally.

**Comment by tstuefe** [comment]
> > > As I wrote before, I don't see a reason to resolve mallinfo dynamically.
> > 
> > 
> > Doesn't statically using mallinfo trigger the warning that we are trying to avoid?
> 
> Yes, that's why I resolve it dynamically. Otherwise we would need to use a GCC pragma to disable deprecation warnings locally.

Okay, I missed that part; I probably would just remove the warning though but this is fine too.

**Comment by tstuefe** [comment]
> > > As I wrote before, I don't see a reason to resolve mallinfo dynamically.
> > 
> > 
> > Doesn't statically using mallinfo trigger the warning that we are trying to avoid?
> 
> Yes, that's why I resolve it dynamically. Otherwise we would need to use a GCC pragma to disable deprecation warnings locally.

How high is the chance that this API will ever be removed for real? Has any API ever been removed from the glibc (serious question)? If the chance is low or zero I'd prefer the runtime loader to resolve my functions for me (so, pragma).

File: src/hotspot/os/linux/os_linux.cpp (hunk 4 above)
Comment (tstuefe) [reply src/hotspot/os/linux/os_linux.cpp:2238]: Which I suggested before. We do this in other places; in fact, every runtime-resolution of a system API via `dlopen` locally hard-codes the prototype of this function - in the form of the function pointer typedef. If that's accepted practice, why are we shy suddenly about making assumptions of how the return structure looks like?

We are not talking about a volatile function but an official glibc interface. mallinfo() did not change in 40 years. I'd expect mallinfo2() to be very stable.

Its not stack allocated. Its pointer is. All we really need to know is offset and size, within this structure, of that one member we access. For that, even copying that whole structure seems overkill IMHO.

File: src/hotspot/os/linux/os_linux.cpp (hunk 5 above)
Comment (tstuefe) [reply src/hotspot/os/linux/os_linux.cpp:174]: Yes, unless you want unused-variable warnings on Alpine.

**Comment by mychris** [comment]
I don't know how glibc handles this, but the release notes of 2.33 states:

> The deprecated <sys/vtimes.h> header and the function vtimes have been removed.

So it looks like they are removing deprecated functions. I guess this doesn't mean that the symbol is no longer in the runtime library (because of symbol versioning, it should remain there), but the function declaration in the header has been removed.

I can confirm, that on a recent Arch Linux (which already has glibc 2.33) this header is no longer to be found in /usr/include. In the libc.so file, however, the symbol is still there:

```
$ readelf --dyn-sym -W /usr/lib/libc.so.6 |grep vtimes
   673: 00000000000f5c40    56 FUNC    GLOBAL DEFAULT   16 vtimes@GLIBC_2.2.5
```

**Comment by tstuefe** [comment]
> I don't know how glibc handles this, but the release notes of 2.33 states:
> 
> > The deprecated <sys/vtimes.h> header and the function vtimes have been removed.
> 
> So it looks like they are removing deprecated functions. I guess this doesn't mean that the symbol is no longer in the runtime library (because of symbol versioning, it should remain there), but the function declaration in the header has been removed.
> 
> I can confirm, that on a recent Arch Linux (which already has glibc 2.33) this header is no longer to be found in /usr/include. In the libc.so file, however, the symbol is still there:
> 
> ```
> $ readelf --dyn-sym -W /usr/lib/libc.so.6 |grep vtimes
>    673: 00000000000f5c40    56 FUNC    GLOBAL DEFAULT   16 vtimes@GLIBC_2.2.5
> ```

You are right :( Okay, then resolving mallinfo manually may be the best option.

File: src/hotspot/os/linux/os_linux.cpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/os/linux/os_linux.cpp:2238]: Sorry I missed the static use of the mallinfo2 struct.

If we want to be able to build and run in any combination then we need to workaround the struct definition.

With the current code if you build on an old system and run on a new one you will still use the old mallinfo. That probably isn't terrible, but assumes our minimal supported build systems will have mallinfo2 long before mallinfo is physically removed.

**Comment by mychris** [comment]
I agree, that for compatibility reasons, we need to copy the struct definition to the VM sources. I did that and guarded the usage with `#ifdef __GLIBC__`. I copied the whole struct definition for both structs, to be forward and backward compatible, into the sources and adjusted the usage accordingly. I also made the declaration of the function pointers similar to others declared in os_linux.hpp.

@tstuefe: I don't know what you mean by:

> Its not stack allocated. Its pointer is. All we really need to know is offset and size, within this structure, of that one member we access. For that, even copying that whole structure seems overkill IMHO.

The return type of `mallinfo` is not a pointer, it is the struct. Did you miss that, or do I missinterpret what you wrote?

**Review by dholmes-ora — APPROVED** [review]
This latest version seems acceptable to me. It avoids the warning and uses a better API when available.

Thanks,
David

**Review by tstuefe — APPROVED** [review]
Looks good to me too. Thank you for fixing this, and sorry for the many iterations.

Cheers, Thomas

**Comment by mychris** [comment]
Sometimes good solutions take some time (even if the problem seems small), so no worries about the iterations. Thank you two a bunch for the reviews and comments.

I don't understand why the Windows build is failing in the CI out of nowhere (it was fine in the first iteration), but from the mailing list discussion I understand that the github hooks don't need to be green. Since the changes clearly don't have an impact in Windows I am ignoring that.

/integrate

**Comment by tstuefe** [comment]
> Sometimes good solutions take some time (even if the problem seems small), so no worries about the iterations. Thank you two a bunch for the reviews and comments.
> 
> I don't understand why the Windows build is failing in the CI out of nowhere (it was fine in the first iteration), but from the mailing list discussion I understand that the github hooks don't need to be green. Since the changes clearly don't have an impact in Windows I am ignoring that.
> 

Yes, that is fine. AFAIK the standing agreement atm is "look at the GHAs, if you think it has nothing to do with you, ignore".

> /integrate
